Success animation for Model R

TychoVrahe opened the pull request:
Imeplements #2293 

Implements also popup screen from design, but text is not centered (limitation of Paragraph implementation in rust, seems out of scope for this PR to implement that). Button in the popup also differs, since out current button implementation does not follow design. Fonts used in design are also not available.

Builds on #2276 which should be merged first.

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -227,6 +748,42 @@ impl Font {
+            return None;
+        }
+
+        unsafe {
```
this whole block should probably be moved to something like `pub unsafe fn Glyph::load(gl_data: *const u8) -> Self`
then maybe we don't even need the `Glyph::new` function, because `load`ing a Glyph is the only supported way to get one

TychoVrahe replied:
done in 3fa8513c9ca6f91afac8465beaec310815863f2c

matejcik reviewed (commented):
partial review so far

Overall, i worry about performance of the new functions, esp. generating the overlay pixel by pixel that incurs glyph lookup per every pixel. This might be good enough for 1bpp Trezor R, but I strongly suspect that it won't be good enough for the T.

Still, probably OK to keep it and worry about performance only when we need to.

TychoVrahe commented on core/embed/rust/src/trezorhal/display.rs:
```diff
@@ -136,3 +143,12 @@ pub fn set_window(x0: u16, y0: u16, x1: u16, y1: u16) {
     }
 }
+
+pub fn get_offset() -> (i32, i32) {
```
Purpose of this is to be consistent with the C functions, what is/was the purpose of the original i don't know, as far as i can tell its not being set anywhere. (function setting the offset is also exposed to python, but even that does not seem to be called anywhere).

jpochyla replied:
IIRC it was for some sort of scrolling / panning animation?

TychoVrahe commented on core/embed/rust/src/trezorhal/ffi.rs:
```diff
@@ -1,5 +1,6 @@
 #![allow(non_camel_case_types)]
 #![allow(non_upper_case_globals)]
+#![allow(non_snake_case)]
```
the uzlib_uncomp struct has a member named 'lzOff'

TychoVrahe commented on core/embed/rust/src/trezorhal/uzlib.rs:
```diff
@@ -0,0 +1,42 @@
+use super::ffi;
+use core::ptr;
```
rustification done in 8e644a6641cb515073e2c43227c33b720ff3b263, though its a bit different as window usage is optional and dest_buf can have arbitrary size

matejcik replied:
any reason `dest_buf` is passed at initialization time instead into the decompression function?

matejcik replied:
do we know what is the purpose of not specifying a window?

TychoVrahe replied:
dest_buf: saves a bit of time by only calculating dest_limit only once maybe? would there by any benefit in using different buffers when iterating?

window: i don't really understand what is happening inside the uzlib, but it seems like the window is not needed when doing full-dest-in-memory

TychoVrahe replied:
one clear advantage of passing `dest_buf` directly to decompressin function is that i can use the data in dest_buf directly instead of copying the return value, which comes handy when using larger buffer that 1 byte. so changed this in https://github.com/trezor/trezor-firmware/commit/a173efb3834e1c28c28aede9e53967b58a0d1b84 and also added some reaction to uzlib error

TychoVrahe commented on core/embed/rust/src/trezorhal/uzlib.rs:
```diff
@@ -0,0 +1,42 @@
+    }
+}
+
+pub fn uzlib_uncompress(decomp: &mut uzlib_uncomp, dest: &mut [u8]) -> i32 {
```
well yes, the fastest is to do whole decompression in one call (=dest buffer is the size of data), like its done in rect_rounded2_partial function exactly for the performance reasons, but for some functionality this is not usable (large icons/images would take too much ram). in rect_rounded2_partial the size of icon is limited to 64*64, so the size is not that critical. If needed (and possible for the given algorithm), we can try using the buffer size somewhere in the middle. What exactly would you like faster?

matejcik replied:
In the past I was unhappy with homescreen decompression speed. That is precisely the case where it won't fit into RAM as a whole, but scanning in chunks could presumably bring speedups.
Also the small black&white icons could certainly be decompressed in one go, RAM should not be an issue there and I see no reason not to do this small optimization.

TychoVrahe replied:
ok i will make some experiments

TychoVrahe replied:
Some results from the testing:
- the fastest is decompress as a whole, without window usage
- slowest is to decompress byte by byte
This is where intuitive results end though.
When decompressing by chunks, fastest seem to be using very small chunks of 2 or 4 bytes, after that, with increasing chunk size the speed actually goes down. I was testing this on two images, the best chunk size differ in each. I also tried comiling the uzlib with -O3 for one image, which sped up the decompressing quite nicely, but didn't change the dependence on the chunk size.
Even when decompressing in one chunk when using window, its way slower than without window usage or small chunks.

Just to have some idea what increases we can get:
Decompressing by best chunk was 17% to 65% faster than per byte
Decompressing by 128 byte chunk was 7% to 40% faster than per byte.
Decompressing whole image without window was 70% to 110% faster than per byte.
This will differ a lot by image. Maybe testing on more images would bring better results. 

So based on this, i would suggest using chunks of 4 bytes for decompression of large images (that means modify display_icon, display_image functions) and implementing new display_small_icon with some reasonable limits (loader using 64*64 so perhaps that) where we could do the whole decompression at once.

matejcik replied:
Curious! Nice research.

I read a bit of the uzlib code. The no-window gains are obvious; essentially, whenever you're using a window, you are copying every byte twice, plus some conditional handling. I suspect we would gain another 2x speedup just by splittting into `uzlib_uncompress_window` and `uzlib_uncompress_nowindow` (honestly that's one of the things `-O3` is probably doing behind the scenes)

The with-window gains are super questionable, because internally the library decodes byte by byte anyway. However, if empirical research tells us that 4-byte chunks are faster, let's use that. (something something CPU word size??)

I'm also wondering if we could hide this inside `display_icon`, by looking at how big we expect the result to be, and choosing the decompression mode? Instead of having a separate `display_small_icon`.

matejcik replied:
...actually, I don't think `display_icon` is ever used for large images. Let's place a hard limit on the icon size and always decompress the full thing; and keep the chunked variant only for `display_image`

TychoVrahe replied:
Largest seems to be 'homescreen_model_r' with 92*92, which is some 4kB of data, with current impl stored on stack - that seems already too much. But apart from this one it looks like we should fit into 64*64 (2kb) which we are already decompressing full in display_loader. 

We would need to convert the homescreen to image format, which will produce larger data stored on flash as it uses full colors (2 byte per pix instead of 4 bits), and also allow display_image function on model R, which we didn't need so far.

The hiding large/small in one function could work, but i imagine we would need to take the stack space for full decompression of smaller icons even when decompressing by small chunks. But that doesn't seem like a big problem.

Also looking at the display_image function: it's already decompressing by chunks of 2 bytes unlike display_icon, so there is probably not that much to gain.

matejcik replied:
aren't icons 1bpp so 64x64 should come out to 512 bytes? (and the homescreen is 1058 bytes)

TychoVrahe replied:
no they are greyscale 4bpp

matejcik replied:
Right.
So, there's the TR homescreen, and we really don't want to upscale that to fullcolor.
And there is the loader icon which is 64x64 (but will it be calling `display_icon`? the decompression is internal)
Other icons are at most 24x24, so that is IMHO the right size limit.

The fact that TR homescreen is rendered via `display_icon` is semantically incorrect, but i don't see anything elegant to do with that for now. We _could_ redefine `display_image` so that on TR it accepts 4bpp images, or we could stuff the feature into `display_icon`, or create a separate function for that purpose -- any other options?

TychoVrahe replied:
We could also implement a B&W icon/image function, modify the data generation to use 1bpp and use larger image size limit there. 

Though i am not sure that setting a hard limit on either of these functions is that great and might need reevaluation when designers desire slightly larger for some reason.

There probably also isn't reason to have the limit strictly squared, using just limit on pixel count will give us a bit more manipulation space.

- loader doesn't call display_icon, i just used that as reference of what we are using elsewhere to estimate what we can(need to be able to) afford on stack
- model 1 homescreen also doesn't fit 24*24, didn't check others so can't confirm nor deny that the rest fits

Anyway, this feels like it should be maybe moved to a separate issue

matejcik replied:
agreed: https://github.com/trezor/trezor-firmware/issues/2414

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -1,11 +1,27 @@
 
 use super::geometry::{Offset, Point, Rect};
 
+pub fn clamp_coords(r: Rect) -> Rect {
```
done in 3fc554d64236e5fa847e0348dfc72f10670f4fb7

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +167,340 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+}
+
+impl TextOverlay {
+    pub fn new(bg_color: Color, fg_color: Color, text: &'static str, font: Font) -> Self {
```
done in cebc7a8865fd3fc689934e065480828b6d650e89

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +167,340 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+pub struct TextOverlay {
+    colortable: [Color; 16],
+    area: Rect,
+    text: &'static str,
```
done in cebc7a8865fd3fc689934e065480828b6d650e89

TychoVrahe commented on core/embed/rust/src/trezorhal/display.rs:
```diff
@@ -136,3 +143,12 @@ pub fn set_window(x0: u16, y0: u16, x1: u16, y1: u16) {
+        let mut x = [0; 1];
+        let mut y = [0; 1];
+        ffi::display_offset(ptr::null_mut(), x.as_mut_ptr(), y.as_mut_ptr());
+        (x[0], y[0])
```
done in 1089016b8104ad479f687795c2e0e8d57da7c1b2, went a little further and returning Offset and using translate function of Rect for the adjustment

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +600,131 @@ pub fn set_window(window: Rect) {
+}
+
+pub struct Glyph {
+    width: i32,
+    height: i32,
+    adv: i32,
+    bearing_x: i32,
+    bearing_y: i32,
```
they could be, but these are mostly used for computations with other geometry structures like Rect where i32 is used anyway (like in get_pixel function of the TextOverlay), so i chose casting here since it makes the actual computations more readable.

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -227,6 +748,42 @@ impl Font {
+            if let Some(gly) = g {
+                let adv = gly.print(baseline + Offset::new(adv_total, 0), colortable);
+                adv_total += adv;
+            }
```
The C `display_get_glyph` function returns nonprintable glyph if it finds that the character is for example unicode, so that glyph is printed. `display_get_glyph` returns null for successive bytes of unicode characters, which we do not want to print. (since in rust the encoding is checked before calling `display_get_glyph` the successive bytes should be discarded before actually calling `display_get_glyph`). Therefore the else branch seems redundant

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +600,131 @@ pub fn set_window(window: Rect) {
+    adv: i32,
+    bearing_x: i32,
+    bearing_y: i32,
+    data: *const u8,
```
well that is nice improvement: 3fa8513c9ca6f91afac8465beaec310815863f2c

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -227,6 +748,42 @@ impl Font {
+
+    pub fn display_text(
+        self,
+        text: &'static str,
```
fixed fe578b3376b06bddc64ed345a5b30267a10da70f

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -1,11 +1,27 @@
+    let x1 = min(r.x1, constant::WIDTH);
+    let y1 = min(r.y1, constant::HEIGHT);
+
+    Rect::new(Point::new(x0, y0), Point::new(x1, y1))
```
seems reasonable to avoid when possible, done in d78bb21ed8c21345b195ee4f9afbb60bc86658eb

TychoVrahe commented:
pushed some fixes requested in review and also:
fixed a off by one bug in a8634321339c29bdf3221cc46e7ff16f4ce3a428
removed confusing comment: 33c770c2048082dec46aab8ee2d41296a5440267
removed unnecessary casting: 60127ff66a5973465d727917a942fd51c83cc590

mmilata commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +600,131 @@ pub fn set_window(window: Rect) {
+        }
+    }
+
+    pub fn get_advance(&self) -> i32 {
```
+1 from me

mmilata reviewed (commented):
This is looking good but will need a few mostly cosmetic changes. Especially please use the higher-level geometric types as much as possible.

Please note that after rebase the animation doesn't work probably because I broke the timer handling in 3dcdffe2032ac710ad3f25b68af2b7edbb39f3a9 and will change it again in #2335. [This patch](https://gist.github.com/mmilata/58554180ea96b02a0b8ed0bcf04e05ba) can be used in the meantime. With this workaround the animation is working fine!

TychoVrahe commented on core/embed/rust/src/trezorhal/uzlib.rs:
```diff
@@ -0,0 +1,64 @@
+            Some([0_u8; UZLIB_WINDOW_SIZE])
+        } else {
+            None
+        };
```
nice, fixed: 92fef13f5c5847438d735aecb8fb5f09a855f6f8

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,339 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+    }
+}
+
+fn get_point_on_line(v0: (i32, i32), v1: (i32, i32), position: f32) -> (i32, i32) {
```
done here b964e7ef103fc64e2c3a2fbc5978030ae67d7e6a , hopefully i didn't miss much. 

i also tried adding and iterator to Rect, which returned directly Point, to eliminate repetitive nested loop iterating: but it came with non negligible performance downgrade, so i left it as it was.

TychoVrahe commented on core/.changelog.d/2293.added:
```diff
@@ -0,0 +1 @@
+Success animation for Model R
```
nothing simpler than deleting it, but in the past i was specifically asked to add log entry for model R by @matejcik - see https://github.com/trezor/trezor-firmware/pull/2243#issuecomment-1121003770 - this seems to be contradicting it

matejcik replied:
I agree with removal here. I wouldn't say "only changes that affect current users", developers are also an audience for the changelog, but it is a matter of granularity. "We had Model R support since here" is good to know; "we had success animation", not so much.

TychoVrahe replied:
removed https://github.com/trezor/trezor-firmware/commit/aa9b5b24f01fb561460af15143fa54153056abe4

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        let mut button_confirmed = false;
+
+        if let Some(b) = self.button.as_mut() {
```
I will have to rebase to be able to use this, which will invalidate the commit hashes in replies here - so i will leave this until everything else is resolved?

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/pull/2339/commits/449cbc278f71b4b9955b4f71a66509f6c08e3379

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,339 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+            let x_t = x - self.area.x0;
+            let y_t = y - self.area.y0;
+
+            for c in self.text.chars() {
```
fixed https://github.com/trezor/trezor-firmware/commit/4f4200a1c16f88f071e619c1da1211bbab395db2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,339 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+    )
+}
+
+fn get_vector(angle: i32) -> (i32, i32) {
```
added https://github.com/trezor/trezor-firmware/commit/27cc200addce41079df2b1133f33ed1c4e8330e9

TychoVrahe commented on core/embed/rust/src/ui/geometry.rs:
```diff
@@ -297,7 +297,32 @@ impl Rect {
+                self.x0
+            } else {
+                limit.x0
+            }, //max
```
fixed https://github.com/trezor/trezor-firmware/commit/b7d318e8b32f817dccf0e13f4b90b01731a7446c

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+impl Component for ResultPopup {
+    type Msg = ResultPopupMessage;
+
+    fn place(&mut self, bounds: Rect) -> Rect {
```
refactored using Rect::split_*: https://github.com/trezor/trezor-firmware/commit/79ba1a004a507e87b9d2bf054b405d475482c189

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+        self.area
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
```
fixed https://github.com/trezor/trezor-firmware/commit/5beba997823aa5729b88460c9686ede19afc6311

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+    state: State,
+    result_anim: Child<ResultAnim>,
+    headline_baseline: Point,
+    headline: Option<&'static str>,
```
fixed https://github.com/trezor/trezor-firmware/commit/5beba997823aa5729b88460c9686ede19afc6311

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+    },
+};
+
+pub enum ResultPopupMessage {
```
fixed https://github.com/trezor/trezor-firmware/commit/5beba997823aa5729b88460c9686ede19afc6311

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+
+    pub fn reset(&mut self, ctx: &mut EventCtx) {
+        self.state = State::Initial;
+        ctx.request_anim_frame();
```
fixed https://github.com/trezor/trezor-firmware/commit/5beba997823aa5729b88460c9686ede19afc6311

The reset function is meant to be used from outside, to restart the popup and the animation, so it is unused in master as nothing uses this yet. 

In this fix i also moved some more logic from `event` to this reset function, and renamed it to `start` - previously the animation started after receiving any event which seemed wrong

TychoVrahe commented on core/embed/rust/src/ui/model_tr/component/result_popup.rs:
```diff
@@ -0,0 +1,211 @@
+pub enum State {
+    Initial,
+    Animating,
+    AnimationDone,
```
fixed https://github.com/trezor/trezor-firmware/commit/5beba997823aa5729b88460c9686ede19afc6311

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,339 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+
+    (
+        ((v0.0 as f32 * rel0 + v1.0 as f32 * rel1) / 2_f32) as i32,
+        ((v0.1 as f32 * rel0 + v1.1 as f32 * rel1) / 2_f32) as i32,
```
done https://github.com/trezor/trezor-firmware/commit/fb70991a8c9100a98ce832e995f1973776dac0a6

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +583,128 @@ pub fn set_window(window: Rect) {
+    let cr: u16 = ((color0.r() as u16) * step + (color1.r() as u16) * (15 - step)) / 15;
+    let cg: u16 = ((color0.g() as u16) * step + (color1.g() as u16) * (15 - step)) / 15;
+    let cb: u16 = ((color0.b() as u16) * step + (color1.b() as u16) * (15 - step)) / 15;
+    Color::rgb(cr as u8, cg as u8, cb as u8)
```
done https://github.com/trezor/trezor-firmware/commit/fb70991a8c9100a98ce832e995f1973776dac0a6

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +583,128 @@ pub fn set_window(window: Rect) {
+                adv: *data.offset(2) as i32,
+                bearing_x: *data.offset(3) as i32,
+                bearing_y: *data.offset(4) as i32,
+                data: slice::from_raw_parts(data.offset(5), (width * height) as usize),
```
fixed https://github.com/trezor/trezor-firmware/commit/70c9a21da588234804504288e7e9047197ff6e29

mmilata commented on core/embed/rust/src/trezorhal/uzlib.rs:
```diff
@@ -0,0 +1,58 @@
+        ctx
+    }
+
+    pub fn uncompress(&mut self, dest_buf: &mut [u8]) -> Result<(), ()> {
```
The C code also returns if there is more data available. Maybe this could return something like `Option<bool>`?

TychoVrahe replied:
used Result<bool, ()> https://github.com/trezor/trezor-firmware/commit/c6e8aafa987aa301536ae4561ca68689415e128a

mmilata reviewed (approved):
Looking good, left some comments but otherwise don't see any issues.

`ui/display.rs` is growing quite large and it would be nice to split it into multiple files but we can do it in another PR since this one is pretty large already.

TychoVrahe commented on core/embed/rust/src/trezorhal/display.rs:
```diff
@@ -1,5 +1,7 @@
 use super::ffi;
+use crate::ui::geometry::Offset;
```
added a wrapper: https://github.com/trezor/trezor-firmware/commit/1a78e3243a9fd6896a122fdb4b016f1401f9010a

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -91,6 +93,52 @@ pub fn icon(center: Point, data: &[u8], fg_color: Color, bg_color: Color) {
+                    if let Ok(()) = ctx.uncompress(&mut dest) {
+                    } else {
+                        return;
+                    }
```
fixed https://github.com/trezor/trezor-firmware/commit/f6685c72531b0ba866129339c4cc4e0041dbe78d

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +153,345 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+        270..=314 => Point::lerp(v[6], v[7], (angle - 270) as f32 / 45_f32),
+        315..=359 => Point::lerp(v[7], v[0], (angle - 315) as f32 / 45_f32),
+        _ => Point::new(1000, 0),
+    }
```
Using your suggestion: https://github.com/trezor/trezor-firmware/commit/e18e0fbdf4ded6fc452cedd84cfe119a5bf53d99

about sin/con: one thing is that we don't satisfy `#[lang = "f32_runtime"]` so we don't have these available (and i don't know how to enable this and what other implications there might be if any). Then we would need to evaluate flash usage and execution time. But if we are gonna do some more animations and graphics, these might come handy, so it would be worth experimenting with at least.

TychoVrahe replied:
Tried it with sin/cos: we don't have these function in rust since they are part of std, which we don't use. Tried using rust crate libm, which worked but it produced about 7kB larger binary. Tried also compiling&exposing C sin/cos functions, which is a bit better - about 4kB more. Didn't check the performance as the size is imo blocker until we really need it.

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+        self.area = area;
+    }
+
+    pub fn get_pixel(&self, underlying: Option<Color>, p: Point) -> Option<Color> {
```
it seems this should always return a `Color` instead of an Option. if the pixel is out of bounds, the underlying color is returned.

also I don't understand the usecase for not specifying an underlying color?

TychoVrahe replied:
changed: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

I used the unspecified underlying in the hold to confirm animation - where i didn't want to specify it, as in the caller the color was used only as info whether to negate the color (if the text color differed from underlying). Changed it - removed the colortable from overlay struct and this function gets both underlying and FG color for the text from the caller. This way it can support more usecases (i.e. deciding what color the text should be based on luminance of the underlying - which should be done in the caller function).

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+/// radius 2
+pub fn rect_rounded2_get_pixel(
+    p: Offset,
+    w: i32,
```
`w` and `h` should probably also be passed as an `Offset`, you can call `rect.size()` when passing the argument

TychoVrahe replied:
done: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+
+/// Gets a color of a pixel on `p` coordinates of rounded rectangle with corner
+/// radius 2
+pub fn rect_rounded2_get_pixel(
```
does this need to be `pub` ?

TychoVrahe replied:
it does not: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+            let underlying_color =
+                rect_rounded2_get_pixel(r_offset, r.width(), r.height(), colortable, filled, 1);
+
+            let mut overlay_color = None;
+            if let Some(o) = overlay {
+                overlay_color = o.get_pixel(None, p);
+            }
+
+            let mut final_color = underlying_color;
+
+            if let Some(overlay) = overlay_color {
+                if overlay == fg_color {
```
```
let overlay_color = overlay.map_or(underlying_color, |o| o.get_pixel(underlying_color, p))
let final_color = if (overlay_color == fg_color) {
    underlying_color.negative()
} else {
    underlying_color
}
```

TychoVrahe replied:
done https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

(differs a bit as the `get_pixel` now accepts the text fg color from here)

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+                continue;
+            }
+
+            if !icon_pixel {
```
it seems icon_pixel must be false here because of the above condition?

TychoVrahe replied:
true: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+    let mut icon_data = [0_u8; ((MAX_ICON_SIZE * MAX_ICON_SIZE) / 2) as usize];
+    let mut icon_width = 0;
+
+    if let Some(i) = icon {
```
fwiw you can write `if let Some((icon_bytes, icon_color)) = icon`

TychoVrahe replied:
seems nicer: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

matejcik commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -204,6 +590,113 @@ pub fn set_window(window: Rect) {
+    /// This function is unsafe because the caller has to guarantee that `data`
+    /// is pointing to a memory containing a valid glyph data, that is:
+    /// - contains valid glyph metadata
+    /// - data has appropriate size
```
also, `data` must actually have a static lifetime

TychoVrahe replied:
added: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -91,6 +93,50 @@ pub fn icon(center: Point, data: &[u8], fg_color: Color, bg_color: Color) {
+
+            if clamped.contains(p) {
+                if x % 2 == 0 {
+                    if ctx.uncompress(&mut dest).is_err() {
```
changed https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

(done everything here in a single fixup, hopefully it isn't too annoying to check, but given that most of these issues are very simple, it felt more appropriate)

matejcik replied:
> done everything here in a single fixup

i don't care review-wise, but this will seriously bite you because you can't autosquash anymore :(

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+
+                        if overlay_data > 0 {
+                            if let Some(u) = underlying {
+                                overlay_color = Some(Color::lerp(
```
changed: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+    pub fn get_pixel(&self, underlying: Option<Color>, p: Point) -> Option<Color> {
+        let mut overlay_color = None;
+
+        if self.area.contains(p) {
```
done: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+
+            for c in self.text.bytes() {
+                if let Some(g) = self.font.get_glyph(c) {
+                    let w = g.width;
```
done: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+
+            let p_rel = Point::new(p.x - self.area.x0, p.y - self.area.y0);
+
+            for c in self.text.bytes() {
```
done: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

TychoVrahe commented on core/embed/rust/src/ui/display.rs:
```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+                        Point::new(tot_adv + b_x + w, b_y),
+                    );
+
+                    if char_area.contains(p_rel) {
```
done: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2

matejcik commented:
looks good from my side. please rebase, squash as appropriate, and implement https://github.com/trezor/trezor-firmware/pull/2339#discussion_r923900217

matejcik commented on core/embed/rust/src/trezorhal/uzlib.rs:
```diff
@@ -0,0 +1,58 @@
+
+impl<'a> UzlibContext<'a> {
+    pub fn new(src: &'a [u8], use_window: bool) -> Self {
+        let window = use_window.then(|| [0_u8; UZLIB_WINDOW_SIZE]);
```
clippy complains that you should use `use_window.then_some([0u8...])`
i'm not entirely convinced due to stack usage but the Option is pre-sized anyway so 🤷‍♀️

TychoVrahe replied:
what clippy version do you use? i don't see it with 0.1.60

TychoVrahe replied:
fixed https://github.com/trezor/trezor-firmware/pull/2339/commits/7d89ad6b173b4850e9871574132886ee2807c88f

matejcik replied:
hmm except our Rust does not know this :( https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/2849519138
i'll go update Rust to latest stable, that should fix the problem

matejcik replied:
no, wait, except it's supposed to be stable for like 8 versions already o_O

TychoVrahe commented:
Squashed, rebased, fixed https://github.com/trezor/trezor-firmware/pull/2339#discussion_r923900217 

Also changed the animation end detection in https://github.com/trezor/trezor-firmware/commit/e32aaf6543194d0ad8458550a78b5fb8a40d516e and https://github.com/trezor/trezor-firmware/commit/413ac64563d9945ad1f02093e05f8e0f7eefe5fd since the issue has been fixed in meantime.

Open points: 
- offset handling - do we want to remove it? maybe do evaluation and possible removal in another issue?
- recall @matejcik not being happy about `Loader` component name for use it the hold to confirm animation, any suggestions?
- not happy about `icon_rust` name also. do we even want to keep this? (it has an advantage of being capable of drawing on odd x coordinate, but mainly i just used it as stepping stone for implementation of more complex display functions)

matejcik commented:
> * offset handling - do we want to remove it? maybe do evaluation and possible removal in another issue?

Let's keep it for now and create an issue.
One thing I would love to try is to scroll off the window contents via the offset functionality.

> * recall @matejcik not being happy about `Loader` component name for use it the hold to confirm animation, any suggestions?

That is true. How about `progress_button` or `button_progress` ?
But we can do that separately too, probably in the same step where we introduce a proper "loader" for TR.

> * not happy about `icon_rust` name also. do we even want to keep this? (it has an advantage of being capable of drawing on odd x coordinate, but mainly i just used it as stepping stone for implementation of more complex display functions)

Let's keep it for now. I would like to do more in this direction, in particular, introduce a `struct Icon`, we can modify the `icon_rust` name in that step.
(maybe @grdddj already did something about a separate Icon type?)

grdddj commented:
> I would like to do more in this direction, in particular, introduce a `struct Icon`, we can modify the `icon_rust` name in that step.
> (maybe @grdddj already did something about a separate Icon type?)

Yes, I am already using `Icon` struct as a convenient way to allow drawing/helper methods on the data (and also to name the icon for debugging purposes).
Current situation:
- `Icon` is defined at https://github.com/trezor/trezor-firmware/blob/grdddj/trezor_r_passphrase_input/core/embed/rust/src/ui/display.rs#L1088-L1164
- individual icons are created as `const` at https://github.com/trezor/trezor-firmware/blob/grdddj/trezor_r_passphrase_input/core/embed/rust/src/ui/model_tr/theme.rs#L19-L53

grdddj commented:
It could be also relevant here - I have made a `HoldToConfirm::icon` and `Loader::icon` constructors, to enable me to create the `HTC` component even with icon data. And I assume we will want the possibility to use `HTC` with icon, the implementation should not be too different than the text one
